### PR TITLE
opened up apt API for extension

### PR DESCRIPTION
**Pull request by lpandzic**

### CODE REVIEW

**Comment by jwgmeligmeyling** [comment]
Interesting, I'm also working on a few QueryDSL APT extensions. Whats your use case?

**Comment by lpandzic** [comment]
Generating Querydsl SQL query classes from Spring Data JDBC model - https://github.com/infobip/infobip-spring-data-querydsl/pull/9

File: querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java [hunk 1]
```diff
@@ -610,6 +617,11 @@ private void serialize(Serializer serializer, Collection<EntityType> models) {
         }
     }
 
+    protected String getClassName(EntityType model) {
+        Type type = conf.getTypeMappings().getPathType(model, model, true);
+        String packageName = type.getPackageName();
+        return !packageName.isEmpty() ? (packageName + "." + type.getSimpleName()) : type.getSimpleName();
```
Comment (kamko) [inline querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java]: Why negated? couldn't you just flip the branches?

File: querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java [hunk 2]
```diff
@@ -401,7 +401,11 @@ private Type createClassType(DeclaredType declaredType, TypeElement typeElement,
         return type;
     }
 
-    private Type createMapType(Iterator<? extends TypeMirror> typeMirrors, boolean deep) {
+    public boolean isSimpleTypeEntity(TypeElement typeElement, Class<? extends Annotation> entityAnn) {
```
Comment (kamko) [inline querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java]: where does this method name come from? (might be ok i just dont see the origin/context)

File: querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java (hunk 1 above)
Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java]: it's cut and paste from 570-572

Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java]: but sure, I'll flip them

Comment (kamko) [reply querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java]: ahh, didn't catch that, but still it looks nicer if flipped

Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java]: fixed

File: querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java (hunk 2 above)
Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java]: line 357, Spring Data JDBC only has one required annotation and that is @Id which is placed on id column and not on the entity class itself hence this change is needed.

File: querydsl-apt/src/main/java/com/querydsl/apt/TypeElementHandler.java [hunk 3]
```diff
@@ -43,7 +43,7 @@
  * @author tiwe
  *
  */
-final class TypeElementHandler {
+public class TypeElementHandler {
```
Comment (robertandrewbain) [inline querydsl-apt/src/main/java/com/querydsl/apt/TypeElementHandler.java]: Why the change?

File: querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java [hunk 4]
```diff
@@ -34,7 +34,7 @@
  * @author tiwe
  *
  */
-final class ExtendedTypeFactory {
+public class ExtendedTypeFactory {
```
Comment (robertandrewbain) [inline querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java]: Why the change?

File: querydsl-apt/src/main/java/com/querydsl/apt/TypeElementHandler.java (hunk 3 above)
Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/TypeElementHandler.java]: In order to customize it - https://github.com/infobip/infobip-spring-data-querydsl/pull/9/files#diff-3ecd73bd5e0c2a2d7e3360d04428617f
I need a hybrid approach between what sql maven plugin and jpa annotation processor do.

File: querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java (hunk 4 above)
Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/ExtendedTypeFactory.java]: Similar to previous comment - https://github.com/infobip/infobip-spring-data-querydsl/pull/9/files#diff-e45095386f2afae8a4499b6f48a71368

File: querydsl-apt/src/main/java/com/querydsl/apt/TypeElementHandler.java (hunk 3 above)
Comment (lpandzic) [reply querydsl-apt/src/main/java/com/querydsl/apt/TypeElementHandler.java]: The only alternative would be to duplicate 80% of the code.
